docs/Software Engineering Blog/posts/spring-boot-3-1-docker-compose-support-intuitive-consistent-develop-run-tests-containers.md

```diff
@@ -0,0 +1,229 @@
+---
+date:
+  created: 2024-09-01
+authors:
+  - NoahHsu
+categories:
+  - Spring Boot
+tags:
+- Java
+  - Development
+  - Unit Test
+  - Integration Test
+  - Docker
+---
+
+# Spring Boot 3.1 Docker Compose Support: The Most Intuitive and Consistent Way to Develop and Run Tests with Containers
+
+![SB-Docker-cover.png](resources%2Fsb-docker%2FSB-Docker-cover.png)
+
+## As-Is
+
+### In Development
+
+Developers often rely on Docker to create a consistent and reproducible environment when developing Spring Boot
+applications. When the number of dependency components grows, two key tools used in this process are Docker Compose and
+Makefile.
+
+However, in a multi-module project, each application might need different dependencies. It comes to a difficult
+situation.
+Every time we need to put [different profiles](https://docs.docker.com/compose/profiles/) with `docker-compose up`, or
+wrap it into [different `make` commands](https://makefiletutorial.com/).
+Both need extra effort to execute and remember (engineers are extremely lazy).
+Don't even mention the integration difficulty in unit tests.
+
+### In Testing
+
+When a test needs to cover the dependency, it first comes up with the mock framework in a programmatic way.
+But it turns out to lose so many details because of the lack of real integration. such as
+
+1. The dialect and SQL execution difference between in-memory h2 DB and a real Mysql DB.
+2. Provide a mock Java object to make an API Client test lose the coverage of the JSON parsing process.
+
+so some of them start to use the TestContainer technic, to provide a real instance during unit tests. However,
+TestContainer suffers from some points too, like:
+
+1. need [support library](https://java.testcontainers.org/modules/databases/mysql/) to gain more interaction.
+2. a programmatic way to set up, which is different from the development time (docker-compose way).
+3. extra setting to share containers in different tests
+
+## Introducing the Spring-Boot-Docker-Compose
+
+In conclusion, the way of providing dependency containers is not a flawless practice for now. So Spring community
```

Comment:
📝 **[vale]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>[Google.Contractions] Use 'isn't' instead of 'is not'.